cmake/CudaRdcUtils.cmake

```
@@ -141,6 +147,20 @@ define_property(TARGET PROPERTY CUDA_RDC_OBJECT_LIBRARY
   FULL_DOCS "Name of the object (without nvlink step) library corresponding to this cuda library"
 )
 
+if(CMAKE_SYSTEM_NAME STREQUAL "Linux")
+  # This is necessary on platform that default to calling ld with --as-needed
+  # (we are looking at you Ubuntu) but also in case user as explicitly using it.
+  set(CMAKE_CXX_LINK_LIBRARY_USING_rdc_no_as_needed_SUPPORTED TRUE CACHE INTERNAL "")
+  set(CMAKE_CXX_LINK_LIBRARY_USING_rdc_no_as_needed
```

Can you add documentation linking to https://cmake.org/cmake/help/latest/variable/CMAKE_LINK_LIBRARY_USING_FEATURE.html ?This also bumps the minimum CMake version for linux+cuda+vecgeom to 3.24.

---

Fair.  Note that we had already broken with CMake v3.18 with:
```
CMake Error at cmake/CudaRdcUtils.cmake:951 (get_target_property):
  INTERFACE_LIBRARY targets may only have whitelisted properties.  The
  property "CUDA_RDC_CACHED_LIB_DEPENDENCIES" is not allowed.
```
I think if useful, I can make it go with:
* Use only if the new code for CMake >= v3.24
* Fail explicitly on Ubuntu if CMake < v3.24

Should I do that or increase the CMake requirement (I will already increase it to v3.19)

---

Um honestly I think the best thing would be a "cmake minimum" at the top of the file. Since:
> CMake also manages a new entry for scripts loaded by [include()](https://cmake.org/cmake/help/latest/command/include.html#command:include) and [find_package()](https://cmake.org/cmake/help/latest/command/find_package.html#command:find_package) commands except when invoked with the NO_POLICY_SCOPE option (see also policy [CMP0011](https://cmake.org/cmake/help/latest/policy/CMP0011.html#policy:CMP0011))

then we could probably just put `cmake_policy(VERSION 3.24...3.31)` at the top of the file?

---

> then we could probably just put cmake_policy(VERSION 3.24...3.31) at the top of the file?

I will add it and it will fail with:

```
CMake Error at cmake/CudaRdcUtils.cmake:127 (cmake_policy):
  An attempt was made to set the policy version of CMake to "3.24" which is
  greater than this version of CMake.  This is not allowed because the
  greater version may have new policies not known to this CMake.  You may
  need a newer CMake version to build this project.
```

Also of notes:
* On Ubuntu 22, the default CMake version is 3.22.1
* On Ubuntu 24 it is 3.28.3
* On the original Spack/Ubuntu 22 failing case, the CMake version is 3.31.4

---

I think that's a decent failure mode.